sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithFailures.java

```diff
@@ -92,6 +93,24 @@
     }
   }
 
+  /**
+   * A handler that holds onto the {@link Throwable} that led to the exception, returning it along
+   * with the original value as a {@link KV}.
+   *
+   * <p>Extends {@link SimpleFunction} so that full type information is captured. {@link KV} and
+   * {@link ComparableThrowable} coders can be easily inferred by Beam, so coder inference can be
+   * successfully applied if the consuming transform passes type information to the failure
+   * collection's {@link TupleTag}. This may require creating an instance of an anonymous inherited
+   * class rather than of this class directly.
+   */
+  public static class ThrowableHandler<T>
+      extends SimpleFunction<ExceptionElement<T>, KV<T, ComparableThrowable>> {
+    @Override
+    public KV<T, ComparableThrowable> apply(ExceptionElement<T> f) {
+      return KV.of(f.element(), ComparableThrowable.forThrowable(f.exception()));
```

Comment:
Unless I missed something, this output will be inferred to use `KvCoder.of(<T coder>. SerializableCoder)`. So this may create problems for pipeline update.

Reply:
Yes, it's using `SerializableCoder`. Should there be a custom coder to handle this, or should it be explicitly registered?

Sorry, I'm not that familiar with how pipeline updates can be affected by changes like this.

Reply:
Pipeline update checks if all coders are the same. Technically only coders used for persisted state in a shuffle. So if the binary encoding of the coder changes, it will be rejected for update. It was my mistake - it is actually `CustomCoder` that may have spurious incompatibilities. For `SerializableCoder` there may be similar risk because it contains the serialized `Class` object.

Reply:
Ack.